### DISCUSSION TITLE
Allow flows scripts to be imported from a file within the project folder

### DISCUSSION THREAD

**Discussion by drennvinn** [directus/directus#23593] [Feature Requests]
### Summary

I propose adding a feature to allow scripts in Directus flows to be imported from a file within the project folder. This would enable versioning of the scripts' content, rather than storing them solely in the database.

### Motivation

Currently, scripts within flows are stored in the database, which limits the ability to version control them and track changes over time. 
By allowing these scripts to be stored as files in the project directory, users can easily manage them through version control systems. This would support a better workflow if you have a big project with a lot of flows that could be complex, long, etc...
This could also give the advantage of using the same file for multiple flows script

### Detailed Design

The proposed feature would introduce an option for users to define scripts in flows using external files located within the project folder. Here’s how it would work:

1. **Script Location**:  
   Users could write a script in the "Run script" form field (current behavior) or imported from a specific file path within the project directory.   
   
![script_location](https://github.com/user-attachments/assets/29383ac9-465e-4460-8084-ed046f075ac2)

2. **Versioning and File Management**:  
   These scripts would be .js (or maybe .ts) stored in a designated folder, such as `flows/my_custom_flow_script.js` or `flows/my-super-flow/my_custom_flow_script.js`. This would allow users to version the scripts alongside their project files. Maybe the system would reference the file at runtime when executing the flow.

3. **UI Integration**:  
   In the Directus UI, when configuring a flow, there could be an option to select whether the script is "database-stored" or "file-based". For file-based scripts, the user would be prompted to enter the file path relative to the project root or simply import the file.
      Here is an example that I quickly made:
      
![Capture d’écran 2024-09-06 à 12 26 03](https://github.com/user-attachments/assets/419f2628-d320-4082-a853-914e054c4708)

### Requirements List

**Must Have**:
  - Support for importing scripts from files located in the project folder.
  - UI option to select between form field for database-stored and file-based script.
  - Error handling for missing or unreadable script files.
  
**Should Have**:
  - Synchronization option to convert database-stored scripts into file-based scripts and vice-versa.
  - Documented guidelines on how to manage scripts through file-based storage, including version control best practices.

### Drawbacks

One potential drawback is increased complexity in managing flow scripts. Storing scripts in files adds an extra layer of setup and responsibility on the developer to ensure that the correct file paths are maintained, and that version control workflows are properly followed.

Additionally, reading from file-based scripts may introduce slight performance overhead compared to fetching scripts directly from the database, though this impact should be minimal.

### Alternatives

...

### Adoption Strategy

This feature should be backward compatible, as it introduces an optional way to manage flow scripts. Existing flows with database-stored scripts would continue to function as they do now, with the file-based option being available as an additional feature.

To assist in the adoption of this feature, documentation would need to be updated, outlining the steps for switching from database-stored scripts to file-based scripts and explaining the benefits of version control integration.

Since this is not a breaking change, it should be easy for developers to adopt this new approach without affecting existing projects.

**Comment by domdus** [comment]
Maybe, aren't [Custom Operations](https://docs.directus.io/extensions/operations.html) the thing you're looking for?

It's basically the same as Run Script, but appears as separate directus operation (build & compiled as isolated script)
+ You can optionally use external npm packages. 
It requires a little dive into how things done. Check out this [guide](https://docs.directus.io/guides/extensions/operations-npm-package.html).

In the end you have your own npm module for version control:
- package.json
- src/app.js
- src/api.js file

**Reply by drennvinn** [reply]
nice! I didn't know about this feature 😅
so I'm going to close this discussion
